File: KIJUNG/PRO_WEEK_06/경주로건설.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int visited[25][25];
int stored[25][25];
int dcol[5] = {0,0,1,0,-1};
int drow[5] = {0,1,0,-1,0};
int answer = 2000000000;

void dfs(int row, int col, int direction, int cost, vector<vector<int>> &board){
    if(row == board.size()-1 && col == board[0].size()-1){
        answer = min(cost,answer);
        return;
    }
    for(int i=1;i<5;i++){
        int nCost = 100;
        if(row+drow[i]<0 || row+drow[i]>board.size()-1 || col+dcol[i]<0 || col+dcol[i]>board[0].size()-1) continue;
        if(visited[row+drow[i]][col+dcol[i]] == 1) continue;
        if(board[row+drow[i]][col+dcol[i]] == 1) continue;
        if((direction == 1 && i%2 == 0) || (direction == 0 && i%2 == 1)) nCost += 500;
        if(stored[row+drow[i]][col+dcol[i]] + 300 < cost+nCost) continue;
        stored[row+drow[i]][col+dcol[i]] = cost+nCost;
        visited[row+drow[i]][col+dcol[i]] = 1;
        dfs(row+drow[i],col+dcol[i],i%2,cost+nCost,board);
        visited[row+drow[i]][col+dcol[i]] = 0;
    }
}

int solution(vector<vector<int>> board) {
    fill(&visited[0][0], &visited[24][25], 0);
    fill(&stored[0][0], &stored[24][25], 2000000000);
    stored[0][0]=0;
    visited[0][0]=1;
    dfs(0,0,2,0,board);
    return answer;
}

File: DONGUK/프로그래머스_WEEK 8/경주로 건설.py
from collections import deque
import math

INF = math.inf

def search(x, y, board):
    N = len(board)
    
    dp = [[[INF] * N for _ in range(N)] for _ in range(4)]
    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]
    
    q = deque()
    
    for z in range(4):
        dp[z][0][0] = 0
    
    # (x, y, cost, dir)
    q.append((0, 0, 0, -1))
    
    while q:
        top = q.popleft()
        x, y, cost, direction = top[0], top[1], top[2], top[3]
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0 <= nx < N and 0 <= ny < N:
                if board[nx][ny] == 0:
                    n_cost = cost
                    if direction == -1:
                        n_cost += 100
                    
                    elif i == direction:
                        n_cost += 100
                        
                    else:
                        n_cost += 600
                    
                    if n_cost < dp[i][nx][ny]:
                        q.append((nx, ny, n_cost, i))
                        dp[i][nx][ny] = n_cost
    
    
    return min(dp[0][-1][-1], dp[1][-1][-1], dp[2][-1][-1], dp[3][-1][-1])
    
    
def solution(board):
    answer = 0
    
    answer = search(0, 0, board)
    return answer